lib/rack/utils.rb

```diff
@@ -440,24 +440,24 @@ def to_hash
       end
 
       def [](k)
-        super(k) || super(@names[k.downcase])
+        super(k) || super(names[k.downcase])
       end
 
       def []=(k, v)
         canonical = k.downcase.freeze
-        delete k if @names[canonical] && @names[canonical] != k # .delete is expensive, don't invoke it unless necessary
-        @names[canonical] = k
+        delete k if names[canonical] && names[canonical] != k # .delete is expensive, don't invoke it unless necessary
```

Comment:
@tenderlove somehow `@names` is being set to `nil` before the Psych parser "revives" the hash. I think I misunderstood the root cause initially. I'm not really sure when/how `@names` is going nil.

Reply:
When the YAML gets dumped, does the YAML text contain the values in the `names` ivar?  If not, then it sounds like a problem with the way it gets emitted to YAML.

Reply:
Yes, it looks like the YAML text does contain the `names` ivar:

```
a = Rack::Utils::HeaderHash.new("foo" => "bar")
  => {"foo"=>"bar"} 
dump = YAML.dump(a)
  => "--- !ruby/hash-with-ivars:Rack::Utils::HeaderHash\nelements:\n  foo: bar\nivars:\n  :@names:\n    foo: foo\n"
b = YAML.load(dump)
Traceback (most recent call last):
       16: from /Users/jordan/.rvm/rubies/ruby-2.5.1/lib/ruby/2.5.0/psych/visitors/to_ruby.rb:32:in `accept'
       15: from /Users/jordan/.rvm/rubies/ruby-2.5.1/lib/ruby/2.5.0/psych/visitors/visitor.rb:6:in `accept'
       14: from /Users/jordan/.rvm/rubies/ruby-2.5.1/lib/ruby/2.5.0/psych/visitors/visitor.rb:16:in `visit'
       13: from /Users/jordan/.rvm/rubies/ruby-2.5.1/lib/ruby/2.5.0/psych/visitors/to_ruby.rb:311:in `visit_Psych_Nodes_Document'
       12: from /Users/jordan/.rvm/rubies/ruby-2.5.1/lib/ruby/2.5.0/psych/visitors/to_ruby.rb:32:in `accept'
       11: from /Users/jordan/.rvm/rubies/ruby-2.5.1/lib/ruby/2.5.0/psych/visitors/visitor.rb:6:in `accept'
       10: from /Users/jordan/.rvm/rubies/ruby-2.5.1/lib/ruby/2.5.0/psych/visitors/visitor.rb:16:in `visit'
        9: from /Users/jordan/.rvm/rubies/ruby-2.5.1/lib/ruby/2.5.0/psych/visitors/to_ruby.rb:268:in `visit_Psych_Nodes_Mapping'
        8: from /Users/jordan/.rvm/rubies/ruby-2.5.1/lib/ruby/2.5.0/psych/visitors/to_ruby.rb:268:in `each_slice'
        7: from /Users/jordan/.rvm/rubies/ruby-2.5.1/lib/ruby/2.5.0/psych/visitors/to_ruby.rb:268:in `each'
        6: from /Users/jordan/.rvm/rubies/ruby-2.5.1/lib/ruby/2.5.0/psych/visitors/to_ruby.rb:271:in `block in visit_Psych_Nodes_Mapping'
        5: from /Users/jordan/.rvm/rubies/ruby-2.5.1/lib/ruby/2.5.0/psych/visitors/to_ruby.rb:336:in `revive_hash'
        4: from /Users/jordan/.rvm/rubies/ruby-2.5.1/lib/ruby/2.5.0/psych/visitors/to_ruby.rb:336:in `each_slice'
        3: from /Users/jordan/.rvm/rubies/ruby-2.5.1/lib/ruby/2.5.0/psych/visitors/to_ruby.rb:336:in `each'
        2: from /Users/jordan/.rvm/rubies/ruby-2.5.1/lib/ruby/2.5.0/psych/visitors/to_ruby.rb:362:in `block in revive_hash'
        1: from /Users/jordan/.rvm/gems/ruby-2.5.1/gems/rack-2.0.6/lib/rack/utils.rb:471:in `[]='
NoMethodError (undefined method `[]' for nil:NilClass)

Reply:
@tenderlove ah ok so YAML parsers do not `initialize` objects, they call `allocate` them. My guess is that the parser is setting the hash key/value pairs before the ivars are set. Does that sound likely?

Reply:
@tenderlove I confirmed that Psych sets the `elements` before setting `ivars` on the new instance. I think this is probably a Psych issue then. I suppose either `ivars` should be emitted before `elements` or the parser should make sure to process `ivars` before `elements`.

See: https://github.com/ruby/psych/blob/63c94cf4f1a877a84c9bbed2ccabc773f9e56035/lib/psych/visitors/to_ruby.rb#L267

Reply:
Should I close this PR and open an issue on the Psych repo?

Reply:
@jkowens yes, if you can reproduce this outside of Rack, then open an issue on Psych.  I think we just need to copy / paste the header hash class and then remove any methods that have nothing to do with serialization.  Lets keep this issue open until we fix the issue in Psych.

Thanks!